compose/ui/ui/src/webCommonW3C/kotlin/androidx/compose/ui/draganddrop/WebDragAndDropManager.kt

```diff
@@ -112,45 +58,88 @@ internal abstract class WebDragAndDropManager(eventListener: EventTargetListener
         }
     }
 
+    private fun InternalStartTransferScope.startTransfer(dragEvent: DragEvent): Boolean {
+        with (rootDragAndDropNode) {
+            startDragAndDropTransfer(dragEvent.offset) {
+                isActive()
+            }
+
+            if (isActive()) {
+                return acceptTransfer(dragEvent)
+            }
+        }
+
+        dragEvent.preventDefault()
+        return false
+    }
+
+
+    private fun acceptTransfer(dragEvent: DragEvent): Boolean {
+        var acceptedTransfer = false
+
+        with (rootDragAndDropNode) {
+            val evt = DragAndDropEvent(dragEvent.offset, transferData = DragAndDropTransferData(dragEvent.dataTransfer))
+            acceptedTransfer = acceptDragAndDropTransfer(evt)
+
+            if (acceptedTransfer) {
+                onStarted(evt)
+                onEntered(evt)
+            }
+        }
+
+        return acceptedTransfer
+    }
+
     private fun initEvents(eventListener: EventTargetListener, globalEventsListener: EventTargetListener) {
+        var previousDragEventIsStart = false
+
         eventListener.addDisposableEvent("dragstart") { event ->
+            // Both internal (starting from within the application)
+            // and external (triggered by dragging something for the outer world)
+            // trigger the "dragenter" event but we can not set drag image anywhere apart dragstart
+            previousDragEventIsStart = true
             event as DragEvent
 
-            with (rootDragAndDropNode) {
-                startTransferScope.startDragAndDropTransfer(event.offset) {
-                    startTransferScope.dragSessionContext != null
-                }
-
-                if (startTransferScope.dragSessionContext != null) {
-                    val dragEvent = DragAndDropEvent(event.offset)
-                    val acceptedTransfer = acceptDragAndDropTransfer(dragEvent)
-
-                    if (acceptedTransfer) {
-                        onStarted(dragEvent)
-                        onEntered(dragEvent)
+            val scope = InternalStartTransferScope(density)
```

Comment:
Since the density is not constant and can be changed by a user, I think we need to have a density provider here rather than a constant density initialized at the very beginning. 

Not a blocker for this PR, imo, rather a TODO.